### PR TITLE
Port mesh intersection routines from VXL

### PR DESCRIPTION
**Pull request by chetnieter**
Ported several routines from VXL to calculate the closest point and ray intersections with triangles and full meshes. All the tests were ported as well. The ```mesh_intersect``` and ```mesh_closest_point``` routines were renamed to better reflect what those routines are doing. New tests had to be created for those routines since no tests existed for those routines in VXL.

### CODE REVIEW

File: arrows/core/mesh_intersect.cxx [hunk 1]
```diff
@@ -0,0 +1,385 @@
+// This file is part of KWIVER, and is distributed under the
+// OSI-approved BSD 3-Clause License. See top-level LICENSE file or
+// https://github.com/Kitware/kwiver/blob/master/LICENSE for details.
+
+/**
+ * \file
+ * \brief Implementation of mesh operations
+ */
```
Comment (mwoehlke-kitware) [inline arrows/core/mesh_intersect.cxx:8]: Can we use the preferred documentation style for new code, please?
```suggestion
/// \file
/// Implementation of mesh operations.
```

(The guide recommends to omit `\brief` for *member* documentation; not sure we've been doing that for file documentation also, or if AUTOBRIEF works for those, so I'd be okay leaving a `\brief` here. But we should probably add a period and use `///` at least.)

File: arrows/core/mesh_intersect.cxx [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+
+#include "mesh_intersect.h"
+#include <limits>
+#include <vital/logger/logger.h>
```
Comment (mwoehlke-kitware) [inline arrows/core/mesh_intersect.cxx:12]:
```suggestion
#include "mesh_intersect.h"

#include <vital/logger/logger.h>

#include <limits>
```

File: arrows/core/mesh_intersect.cxx [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+
+using namespace kwiver::vital;
+
+/// Intersect the ray from point to a triangle
+bool
+mesh_intersect_triangle( const vital::point_3d& p,
+                         const vital::vector_3d& d,
+                         const vital::point_3d& a,
+                         const vital::point_3d& b,
+                         const vital::point_3d& c,
```
Comment (mwoehlke-kitware) [inline arrows/core/mesh_intersect.cxx:28]: I find it odd to be `using namespace kwiver::vital` but still have `vital::` here (and elsewhere).

Also, if each parameter is going to be on its own line, maybe consider breaking after the `(` so the parameters are only indented one level?

File: arrows/core/mesh_intersect.cxx [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+
+namespace kwiver {
+
+namespace arrows {
+
+namespace core {
+
+using namespace kwiver::vital;
+
+/// Intersect the ray from point to a triangle
```
Comment (mwoehlke-kitware) [inline arrows/core/mesh_intersect.cxx:22]: Separator lines would be nice. Also, duplicating documentation from the header is discouraged.

File: arrows/core/tests/test_mesh_intersect.cxx [hunk 5]
```diff
@@ -0,0 +1,226 @@
+// This file is part of KWIVER, and is distributed under the
+// OSI-approved BSD 3-Clause License. See top-level LICENSE file or
+// https://github.com/Kitware/kwiver/blob/master/LICENSE for details.
+
+#include <gtest/gtest.h>
+
+#include <vital/plugin_loader/plugin_manager.h>
+
+#include <arrows/core/mesh_intersect.h>
+#include <Eigen/Geometry>
```
Comment (mwoehlke-kitware) [inline arrows/core/tests/test_mesh_intersect.cxx:10]:
```suggestion
#include <arrows/core/mesh_intersect.h>

#include <vital/plugin_loader/plugin_manager.h>

#include <Eigen/Geometry>

#include <gtest/gtest.h>
```

**Review by mwoehlke-kitware — CHANGES_REQUESTED** [review]
Not a thorough review, but probably hits the high points assuming the algorithms are correct.

File: arrows/core/mesh_intersect.cxx (hunk 1 above)
Comment (chetnieter) [reply arrows/core/mesh_intersect.cxx:8]: I will make sure to re-read the code style documentation. Is there header/*.cxx that we know has the preferred documentation style.

**Pull request merged**